app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,61 +60,126 @@ abstract class Syncer {
+                remoteCollections[remoteCollection.url] = remoteCollection
+
+        // 2. update/delete local resources and determine new (unknown) remote collections
+        val newRemoteCollections = HashMap(remoteCollections)
+        for (localCollection in localCollections) {
+            val remoteCollection = remoteCollections[getUrl(localCollection)]
+            if (remoteCollection == null)
+                // Collection got deleted on server, delete obsolete local resource
+                delete(localCollection)
+            else {
+                // Collection exists locally, update local resource and don't add it again
+                update(localCollection, remoteCollection)
+                newRemoteCollections -= remoteCollection.url
+            }
+        }
+
+        // 3. create new local collections for newly found remote collections
+        for ((_, collection) in newRemoteCollections)
+            create(collection)
+
+        // 4. sync local resources
+        for (localCollection in localSyncCollections)
+            remoteCollections[getUrl(localCollection)]?.let { remoteCollection ->
+                syncCollection(localCollection, remoteCollection)
+            }
+
+    }
+
+    abstract fun beforeSync()
+
+    abstract fun getUrl(localCollection: CollectionType): HttpUrl?
+
+    abstract fun delete(localCollection: CollectionType)
+
+    abstract fun update(localCollection: CollectionType, remoteCollection: Collection)
+
+    abstract fun create(remoteCollection: Collection)
+
+    abstract fun syncCollection(localCollection: CollectionType, remoteCollection: Collection)
 
     fun onPerformSync(
         account: Account,
         extras: Array<String>,
         authority: String,
         syncResult: SyncResult
     ) {
-        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
-
-        // use contacts provider for address books
-        val contentAuthority =
-            if (authority == context.getString(R.string.address_books_authority))
-                ContactsContract.AUTHORITY
-            else
-                authority
+        // Set sync arguments
+        this.account = account
+        this.extras = extras
+        this.authority = authority
+        this.syncResult = syncResult
 
-        val accountSettings by lazy { accountSettingsFactory.forAccount(account) }
-        val httpClient = lazy { HttpClient.Builder(context, accountSettings).build() }
+        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
 
-        // acquire ContentProviderClient of authority to be synced
-        val provider = try {
-            context.contentResolver.acquireContentProviderClient(contentAuthority)
+        // Acquire ContentProviderClient
+        val tryProvider = try {
+            context.contentResolver.acquireContentProviderClient(authority)
```

Comment:
If possible, I think using `.use` would be better (make it impossible to forget to close).